bluecellulab/circuit_simulation.py

```diff
@@ -330,49 +330,104 @@ def _add_stimuli(self, add_noise_stimuli=False,
         shotnoise_stim_count = 0
         ornstein_uhlenbeck_stim_count = 0
 
+        # Pre-fetch compartment sets (if available) to detect when a stimulus
+        # target refers to a named compartment_set rather than a node_set.
+        compartment_sets: dict[str, dict[str, Any]] | None
+        try:
+            compartment_sets = self.circuit_access.config.get_compartment_sets()
+        except Exception:
+            compartment_sets = None
+
         for stimulus in stimuli_entries:
-            target = stimulus.target
-            gids_of_target = self.circuit_access.get_target_cell_ids(target)
+            # Build a unified list of (cell_id, section, segx, section_name) targets
+            targets: list[tuple] = []
+
+            if stimulus.compartment_set is not None:
+                if compartment_sets is None:
+                    raise ValueError(
+                        "Simulation config provides compartment_set stimuli but "
+                        "no 'compartment_sets_file' is configured."
+                    )
+
+                comp_name = stimulus.compartment_set
+                if comp_name not in compartment_sets:
+                    raise ValueError(f"Compartment set '{comp_name}' not found in compartment_sets file.")
+
+                comp_entry = compartment_sets[comp_name]
+                comp_nodes = comp_entry.get("compartment_set", [])
+
+                # Try to narrow target GIDs using the optional population name
+                population_name = comp_entry.get("population")
+
+                gids_of_target = None
+                if population_name is not None:
+                    try:
+                        gids_of_target = {
```

Comment:
This could be done in the for loop right after this, no? It feels like you are looking at all items in self.cells twice, even though you could look at them only once